pandas/tests/categorical/test_constructors.py

```diff
@@ -464,6 +464,25 @@ def test_from_codes_with_categorical_categories(self):
         with pytest.raises(ValueError):
             Categorical.from_codes([0, 1], Categorical(['a', 'b', 'a']))
 
+    def test_from_codes_with_nan_code(self):
+        # GH21767
+        codes = [1, 2, np.nan]
+        categories = ['a', 'b', 'c']
+        with pytest.raises(ValueError):
+            Categorical.from_codes(codes, categories)
+
+    def test_from_codes_with_float(self):
+        # GH21767
+        codes = [1.0, 2.0, 0]  # integer, but in float dtype
+        categories = ['a', 'b', 'c']
+
+        with tm.assert_produces_warning(FutureWarning, check_stacklevel=False):
```

Comment:
We want to check the stacklevel here. You may need to adjust it in `from_codes` so that the correct line of user-code is reported.